akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala

```diff
@@ -204,6 +204,18 @@ object Source {
     })
 
   /**
+   * Create [[Source]] that will continually produce given elements in specified order.
+   *
+   * Start a new 'cycled' `Source` from the given elements. The producer stream of elements
+   * will continue infinitely by repeating the same sequence of elements provided during construction.
+   */
+  def cycle[T](f: () ⇒ Iterator[T]): Source[T, NotUsed] = {
+    require(f().nonEmpty, "Provided iterator should not be empty")
```

Comment:
Actually giving it another thought: why should stream fail at all? `fromIterator` does not fail the stream in case you pass empty `Iterator` object, stream just finishes. Maybe the same should happen in this case?